stellargraph/core/graph_standard.py

```diff
@@ -0,0 +1,213 @@
+    def _get_node_data(data, kwargs):
+        attrs = {}
+        for name in ["node_id", "node_type", "node_features", "default_node_type"]:
+            value = kwargs.get(name)
+            if value is not None:
+                attrs[name] = value
+        return NodeData(data, **attrs)
+
+    @staticmethod
+    def _get_edge_data(data, kwargs):
+        attrs = {}
+        for name in [
+            "source_id",
+            "target_id",
+            "edge_id",
+            "edge_type",
+            "edge_weight",
+            "default_edge_type",
+            "default_edge_weight",
+        ]:
+            value = kwargs.get(name)
+            if value is not None:
+                attrs[name] = value
+        return EdgeData(data, **attrs)
+
+    ##################################################################
+    # Common interface:
+
+    def is_directed(self) -> bool:
+        return self._is_directed
+
+    def number_of_nodes(self) -> int:
+        return self._node_data.num_nodes()
+
+    def number_of_edges(self) -> int:
+        return self._edge_data.num_edges()
+
+    def nodes(
+        self, include_info: bool = False
+    ) -> Union[Iterable[Any], Iterable[tuple]]:
+        return self._node_data.nodes(include_info)
+
+    def edges(self, include_info: bool = False) -> Iterable[Any]:
+        return self._edge_data.edges(include_info)
+
+    def has_node(self, node: Any) -> bool:
+        return self._node_data.has_node(node)
+
+    def neighbour_nodes(self, node: Any) -> Iterable[Any]:
+        return set(self._edge_cache.neighbour_nodes(node))
+
+    def in_nodes(self, node: Any) -> Iterable[Any]:
+        if self.is_directed():
+            return set(self._edge_cache.in_nodes(node))
+        return self.neighbour_nodes(node)
+
+    def out_nodes(self, node: Any) -> Iterable[Any]:
+        if self.is_directed():
+            return set(self._edge_cache.out_nodes(node))
+        return self.neighbour_nodes(node)
+
+    def node_type(self, node: Any) -> Any:
+        return self._node_data.node_type(node)
+
+    ##################################################################
+    # Computationally intensive methods:
+
+    def info(self, show_attributes: bool = True, sample: Optional[int] = None):
+        directed_str = "Directed" if self.is_directed() else "Undirected"
+        s = "{}: {} multigraph\n".format(type(self).__name__, directed_str)
+        s += " Nodes: {}, Edges: {}\n".format(
+            self.number_of_nodes(), self.number_of_edges()
+        )
+        # TODO Mimic NetworkXStellrGraph output
+        return s
+
+    def node_degrees(self) -> Mapping[Any, int]:
+        raise NotImplementedError
+
+    def adjacency_weights(self):
+        raise NotImplementedError
+
+    ##################################################################
+    # Private methods:
+
+    def check_graph_for_ml(self):
+        # TODO Actually do something!!
+        pass
+
+    def create_graph_schema(
```

Comment:
In the future I think we should move to a pre-calculated schema in the graph class.  How we can do this efficiently is an open question.

Reply:
For example, with edges stored as a dictionary of edge types it will not require parsing all edges to extract the edge types.  Building up an inverse index of edge ID to edge type will also be straightforward.  Similarly for node types.